docker/kuberay-autoscaler/autoscaling_config.py

```diff
@@ -0,0 +1,300 @@
+    resources["CPU"] = num_cpus
+
+    if num_gpus is not None:
+        resources["GPU"] = num_gpus
+
+    if memory is not None:
+        resources["memory"] = memory
+
+    resources.update(custom_resource_dict)
+
+    return resources
+
+
+def _get_num_cpus(
+    ray_start_params: Dict[str, str],
+    k8s_resource_limits: Dict[str, str],
+    group_name: str,
+) -> int:
+    if "num_cpus" in ray_start_params:
+        return int(ray_start_params["num_cpus"])
+    elif "cpu" in k8s_resource_limits:
+        cpu_str = str(k8s_resource_limits["cpu"])
+        if cpu_str[-1] == "m":
+            # For example, '500m' rounds up to 1.
+            return math.ceil(int(cpu_str[:-1]) / 1000)
+        else:
+            return int(cpu_str)
+    else:
+        # Getting the number of CPUs is important, so raise an error if we can't do it.
+        raise ValueError(
+            f"Autoscaler failed to detect `CPU` resources for group {group_name}."
+            "\nSet the `--num-cpus` rayStartParam and/or "
+            "the CPU resource limit for the Ray container."
+        )
+
+
+def _get_memory(
+    ray_start_params: Dict[str, str], k8s_resource_limits: Dict[str, Any]
+) -> Optional[int]:
+    """Get memory resource annotation from ray_start_params, if it is set there.
+
+    TODO, maybe: Consider container resource limits as in
+    https://github.com/ray-project/ray/pull/14567/files
+    """
+    if "memory" in ray_start_params:
+        return int(ray_start_params["memory"])
+    return None
+
+
+def _get_num_gpus(
+    ray_start_params: Dict[str, str],
+    k8s_resource_limits: Dict[str, Any],
+    group_name: str,
+) -> Optional[int]:
+    """Read the number of GPUs from the Ray start params.
+
+    Potential TODO: Read GPU info from the container spec, here and in the
+    Ray Operator.
+    """
+
+    if "num_gpus" in ray_start_params:
+        return int(ray_start_params["num_gpus"])
+
+    # Issue a warning if GPUs are present in the container spec but not in the
+    # ray start params.
+    # TODO: Consider reading GPU info from container spec.
+    for key in k8s_resource_limits:
+        global _GPU_WARNING_LOGGED
+        if "gpu" in key and not _GPU_WARNING_LOGGED:
+            with suppress(Exception):
+                if int(k8s_resource_limits[key]) > 0:
+                    logger.warning(
+                        f"Detected GPUs in container resources for group {group_name}."
+                        "To ensure Ray and the autoscaler are aware of the GPUs,"
+                        " set the `--num-gpus` rayStartParam."
+                    )
+                    _GPU_WARNING_LOGGED = True
+            break
+
+    return None
+
+
+def _get_custom_resources(
+    ray_start_params: Dict[str, Any], group_name: str
+) -> Dict[str, int]:
+    if "resources" not in ray_start_params:
+        return {}
+    resources_string = ray_start_params["resources"]
+    try:
+        resources_json = resources_string[1:-1].replace("\\", "")
```

Comment:
Can you document why the indexing and replacing is needed here? I assume it is related to the unfortunate resource specification below :)

Reply:
Indeed, most unfortunate formatting for the unfortunate resource specification.